tests/testsuite/package.rs

```diff
@@ -5077,3 +5077,573 @@ path = "examples/z.rs"
+            repository = "bar"
+
+            [dependencies]
+            # Let one dependency also specify features, for the added test coverage when generating package files.
+            level2 = { workspace = true, features = ["foo"] }
+        "#,
+            )
+            .file("level1/src/main.rs", "fn main() {}")
+            .file(
+                "level2/Cargo.toml",
+                r#"
+            [package]
+            name = "level2"
+            version = "0.0.1"
+            edition = "2015"
+            authors = []
+            license = "MIT"
+            description = "level2"
+            repository = "bar"
+
+            [features]
+            foo = []
+
+            [dependencies]
+            level3 = { path = "../level3", version = "0.0.1" }
+        "#
+            )
+            .file("level2/src/lib.rs", "")
+            .file(
+                "level3/Cargo.toml",
+                r#"
+            [package]
+            name = "level3"
+            version = "0.0.1"
+            edition = "2015"
+            authors = []
+            license = "MIT"
+            description = "level3"
+            repository = "bar"
+        "#,
+            )
+            .file("level3/src/lib.rs", "")
+            .build();
+
+    p.cargo("package")
+        .replace_crates_io(crates_io.index_url())
+        .with_status(101)
+        .with_stdout("")
+        .with_stderr(
+            "\
+[PACKAGING] level1 v0.0.1 ([CWD]/level1)
+[UPDATING] crates.io index
+error: failed to prepare local package for uploading
+
+Caused by:
+  no matching package named `level2` found
+  location searched: registry `crates-io`
+  required by package `level1 v0.0.1 [..]
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn workspace_with_local_deps_packaging_one_fails() {
+    let crates_io = registry::init();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [workspace]
+            members = ["level1", "level2"]
+        "#,
+        )
+        .file(
+            "level1/Cargo.toml",
+            r#"
+            [package]
+            name = "level1"
+            version = "0.0.1"
+            edition = "2015"
+            authors = []
+            license = "MIT"
+            description = "level1"
+            repository = "bar"
+
+            [dependencies]
+            level2 = { path = "../level2", version = "0.0.1" }
+        "#,
+        )
+        .file("level1/src/lib.rs", "")
+        .file(
+            "level2/Cargo.toml",
+            r#"
+            [package]
+            name = "level2"
+            version = "0.0.1"
+            edition = "2015"
+            authors = []
+            license = "MIT"
+            description = "level2"
+            repository = "bar"
+        "#,
+        )
+        .file("level2/src/lib.rs", "")
+        .build();
+
+    // We can't package just level1, because there's a dependency on level2.
+    p.cargo("package -p level1")
```

Comment:
Should we have a test case when a version-req compatible `level2` is already published?

Reply:
Good point, I added one and it raises a possible issue: a case where `-Zpackage-workspace` fails but old cargo didn't.

If you have a local crate `dep@0.1.0` which is published with the same version `0.1.0`, and you have a crate in the same workspace that depends on `dep@0.1.0` then packaging the workspace used to succeed (because the dependency on `dep` will be satisfied by the registry) but now it fails because the overlay source complains about the version conflict between the published `dep@0.1.0` and the overlaid version...

Reply:
Talked about this in office hour.

Our choices
- Break compatibility on this
- Allow the internal-only source overlap to shadow versions from the original source, rather than error.

`cargo publish` will already fail in this scenario, so the question is strictly about `cargo package` use cases.  crates.io is using `cargo package` but for verifying `.crate` files.  This shouldn't be a problem and we should add a test for this use case.  Multi-package packaging is already fairly busted.  Maybe its ok to break compatibility?

In the end, we should go with one solution, note it in the PR description, and make this an explicit point during the FCP.

Reply:
Remembered the other use case I was wondering about: packaging for the sake of testing the MSRV for registry dependents: https://github.com/taiki-e/cargo-hack/issues/216

Reply:
Makes sense. I think this should also be fine, as again it's per-package instead of per-workspace.

Reply:
I've tweaked the compatibility note